### PR TITLE
support IP adapter SD-XL training & SD-1.5 inference

### PR DESCRIPTION
**Pull request by zhtmike**
TODO:
1. ~~a more close alignment of the sampler~~ 
2. ~~Add image2image pipeline~~ 
3. SD-1.5 training

### CODE REVIEW

File: examples/ip_adapter/assets/woman.png
Comment (SamitHuang) [inline examples/ip_adapter/assets/woman.png:1]: better to upload image via git content url

File: examples/ip_adapter/configs/inference/sd_v15.yaml [hunk 1]
```diff
@@ -0,0 +1,24 @@
+base: ../stable_diffusion_v2/configs/v1-inference.yaml
```
Comment (SamitHuang) [inline examples/ip_adapter/configs/inference/sd_v15.yaml:1]: nice method to integrate sdv2 and sdxl together

File: examples/ip_adapter/adapter/sdv2/modules/attention.py
Comment (SamitHuang) [inline examples/ip_adapter/adapter/sdv2/modules/attention.py:1]: Is this new attention module tested on 910b when flash attention is ON?

File: examples/ip_adapter/README.md [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+    ```bash
+    mpirun -n 8 python train_sdxl.py \
+        --data_path COCO2017/train2017 \
+        --is_parallel True \
+        --weight checkpoints/sd_xl_base_1.0_ms_ip_adapter_train.ckpt
+    ```
+
+    You may also need to increase the training steps and adjust the learning rate in the configuration file to improve the model’s performance.
+
+## Acknowledgements
```
Comment (wtomin) [inline examples/ip_adapter/README.md:229]: Acknowledgements -> Acknowledgments

File: examples/ip_adapter/README.md [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+
+If you want to train the IP-Adapter from scratch, follow these steps:
+
+1. Prepare the checkpoint without IP-Adapter weight by running the following command:
+
+```bash
+python tools/merge_ckpts_sdxl.py --skip_ip --out checkpoints/sdxl_models/merged/sd_xl_base_1.0_ms_ip_adapter_train.ckpt
+```
+
+2. Prepare a relatively large amount of training data for model training, such as LAION-2B and COYO-700M. For illustration purposes, we will use COCO2017 as training data. Download COCO2017 training image, val images and train/val annodations from [https://cocodataset.org/]
```
Comment (wtomin) [inline examples/ip_adapter/README.md:200]: annodations -> annotations

File: examples/ip_adapter/README.md [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+This folder contains [IP-Adapter](https://arxiv.org/abs/2308.06721) models implemented with [MindSpore](https://www.mindspore.cn/), reference to [Official Implementation](https://github.com/tencent-ailab/IP-Adapter).
+
+IP-Adapter is an effective and lightweight adapter to achieve image prompt capability for the pre-trained text-to-image diffusion models. An IP-Adapter with only 22M parameters can achieve comparable or even better performance to a fine-tuned image prompt model. IP-Adapter can be generalized not only to other custom models fine-tuned from the same base model, but also to controllable generation using existing controllable tools. Moreover, the image prompt can also work well with the text prompt to accomplish multimodal image generation.
+
+<p align="center"><img width="700" alt="IP-Adapter Architecture" src="https://github.com/zhtmike/mindone/assets/8342575/f332b980-0fb9-4fe6-bc08-f66c33cabc8f"/>
+<br><em>Overall IP-Adapter architecture</em></p>
+
+## Dependency
+
+- mindspore 2.2
```
Comment (wtomin) [inline examples/ip_adapter/README.md:14]: Only ms 2.2 is supported? Does it mean running on 910A with ms2.1 is not feasible?

File: examples/ip_adapter/assets/woman.png
Comment (zhtmike) [reply examples/ip_adapter/assets/woman.png:1]: changed to url

File: examples/ip_adapter/README.md [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+
+Below are some arguments in the config file that you may want to tune for a better performance on your dataset:
+
+- `per_batch_size`: the number of batch size for training.
+- `base_learning_rate`: the learning rates for training.
+- `total_step`: total number of training steps
+
+For more argument illustration, please run `python train_sdxl.py -h`.
+
+Once you have finished the model fineting, you need to merge the weight from stable diffusion XL, image encoder, and *finetuned* IP-Adapter into a single file, which can be loaded by the inference script later. To merge the checkpoint, please run the following command:
```
[the remark this answers is not in the document]
Comment (zhtmike) [reply examples/ip_adapter/README.md]: fixed. thx

File: examples/ip_adapter/README.md (hunk 2 above)
Comment (zhtmike) [reply examples/ip_adapter/README.md]: fixed. thx

File: examples/ip_adapter/README.md (hunk 3 above)
Comment (zhtmike) [reply examples/ip_adapter/README.md]: fixed. thx

File: examples/ip_adapter/README.md (hunk 4 above)
Comment (zhtmike) [reply examples/ip_adapter/README.md:14]: em. as this model heavily relies on the original implemention of sdxl, as sdxl is only support 2.2 (see [here)](https://github.com/mindspore-lab/mindone/blob/master/examples/stable_diffusion_xl/GETTING_STARTED.md#dependency), so as this model

File: examples/ip_adapter/adapter/sdv2/modules/attention.py
Comment (zhtmike) [reply examples/ip_adapter/adapter/sdv2/modules/attention.py:1]: this attention module is a copy of the original attention module with extra two parameters. So it does not work on 910B, so as the original one currently.

File: examples/ip_adapter/README.md (hunk 4 above)
Comment (wtomin) [reply examples/ip_adapter/README.md:14]: I see. Thanks!

**Comment by wtomin** [comment]
`examples/ip_adapter/assets/.gitkeep` seems unnecessary, which can be deleted.

**Comment by zhtmike** [comment]
> `examples/ip_adapter/assets/.gitkeep` seems unnecessary, which can be deleted.

it is the folder to place the download image (i.e., woman.png), thus we add a `.gitkeep` here to keep the directory `assets` existed

**Comment by wtomin** [comment]
> > `examples/ip_adapter/assets/.gitkeep` seems unnecessary, which can be deleted.
> 
> it is the folder to place the download image (i.e., woman.png), thus we add a `.gitkeep` here to keep the directory `assets` existed

I see. That makes sense.

**Review by wtomin — APPROVED** [review]

**Review by hadipash — APPROVED** [review]

**Review by zhanghuiyao — APPROVED** [review]

**Pull request merged**